gapic/samplegen/samplegen.py

```diff
@@ -86,48 +92,147 @@ class InvalidRequestSetup(SampleError):
+            request (list[dict{str:str}]): The request body from the sample config
+
+        Returns:
+            list[dict{str:(str|list[dict{str:str}])}]: The transformed request block.
+
+        Raises:
+            RedefinedVariable: If an "input_parameter" attempts to redefine a
+                               previously defined variable.
+            ReservedVariableName: If an "input_parameter" value or a "base" value
+                                  is a reserved word.
+            InvalidRequestSetup: If a dict in the request lacks a "field" key
+                                 or the corresponding value is malformed.
+        """
+        base_param_to_attrs: Mapping[str,
+                                     List[Mapping[str, str]]] = defaultdict(list)
+        for field_assignment in request:
+            field_assignment_copy = dict(field_assignment)
+            input_param = field_assignment_copy.get("input_parameter")
+            if input_param:
+                self._handle_lvalue(input_param)
+
+            field = field_assignment_copy.get("field")
+            if not field:
+                raise InvalidRequestSetup(
+                    "No field attribute found in request setup assignment: {}",
+                    field_assignment_copy)
+
+            # TODO: properly handle top level fields
+            # E.g.
+            #
+            #  -field: edition
+            #   comment: The edition of the series.
+            #   value: '123'
+            #   input_parameter: edition
+            m = re.match(r"^([^.]+)\.([^.]+)$", field)
+            if not m:
+                raise InvalidRequestSetup(
+                    "Malformed request attribute description: {}", field)
+
+            base, attr = m.groups()
+            if base in RESERVED_WORDS:
+                raise ReservedVariableName(
+                    "Tried to define '{}', which is a reserved name".format(base))
+
+            field_assignment_copy["field"] = attr
+            base_param_to_attrs[base].append(field_assignment_copy)
+
+        return [TransformedRequest(base, body)
+                for base, body in base_param_to_attrs.items()]
+
+    def validate_response(self, response):
         """Validates a "response" block from a sample config.
 
-        A full description of the response block is outside the scope of this code
-        and is better handled by reading the samplegen documentation.
+        A full description of the response block is outside the scope of this code;
+        refer to the samplegen documentation.
+
+
+        Dispatches statements to sub-validators.
 
         Args:
-           response: list[dict{str:?}]: The structured data representing
-                                        the sample's response.
+            response: list[dict{str:?}]: The structured data representing
+                                         the sample's response.
+
+        Returns:
+            bool: True if response is valid.
```

Comment:
Will this return `False` under some condition in the future? At the moment it either `raise`s or returns `True`

Reply:
That's correct. The pythonic thing to do on an error is raise an exception.
If I were willing to break with the language idiom, I would be returning error objects/None everywhere, with functions that actually provide a return changing return signature to be (Result, Error).

The reason I went with returning `True` instead of not returning anything is that in the unit tests, it feels weird to just...invoke things and say the test passes if no exceptions are raised. It's really just to satiate that feeling: the return value is a complete throwaway.

Reply:
So what you're saying is that it _won't_ return `False` ever? Then I would not return `True` either: either it raises or it returns. I don't think it's that weird in tests; if it bothers, you can add a comment in the tests: `# ensure this completes without exception` .... or a no-op `with assertDoesNotRaise():` ;-)

Reply:
Done